File: engine/src/models/command.rs
use serde::{Deserialize, Serialize};

use super::module::ModuleInfo;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CommandRequest {
    pub command_type: CommandType,
    pub module_name: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CommandType {
    ListModules,
    StartModule,
    StopModule,
    RunExperiment,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CommandResponse {
    pub ok: bool,
    pub command_type: CommandType,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modules: Option<Vec<ModuleInfo>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub module: Option<ModuleInfo>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_path: Option<String>,
}

File: frontend/src/features/terminal/command.ts
export type CommandType =
  | "ListModules"
  | "StartModule"
  | "StopModule"
  | "RunExperiment";

export type ModuleInfo = {
  name: string;
  status: string;
};

export type CommandRequest = {
  commandType: CommandType;
  moduleName?: string;
};

export type CommandResponse = {
  ok: boolean;
  commandType: CommandType;
  message: string;
  modules?: ModuleInfo[];
  module?: ModuleInfo;
  nextPath?: string;
};

export const commandNeedsModuleName = (commandType: CommandType): boolean =>
  commandType === "StartModule" || commandType === "StopModule";

export function buildCommandRequest(
  commandType: CommandType,
  moduleName: string,
): CommandRequest {
  if (!commandNeedsModuleName(commandType)) {
    return { commandType };
  }

  const normalizedName = moduleName.trim();
  if (!normalizedName) {
    throw new Error("module name is required");
  }
  return { commandType, moduleName: normalizedName };
}

export function summarizeCommandResponse(response: CommandResponse): string {
  if (response.module) {
    return `${response.module.name}: ${response.module.status}`;
  }
  if (response.modules && response.modules.length > 0) {
    return response.modules
      .map((module) => `${module.name}: ${module.status}`)
      .join(", ");
  }
  return response.message;
}
